src/couch_replicator/src/couch_replicator_auth_session.erl

```diff
@@ -0,0 +1,486 @@
+]).
+
+-export([
+    init/1,
+    terminate/2,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    code_change/3,
+    format_status/2
+]).
+
+
+-include_lib("ibrowse/include/ibrowse.hrl").
+-include("couch_replicator_api_wrap.hrl").
+
+
+-type headers() :: [{string(), string()}].
+-type code() :: non_neg_integer().
+-type creds() :: {string() | undefined, string() | undefined}.
+
+
+% Behavior API callbacks
+
+-spec initialize(#httpdb{}) -> {ok, #httpdb{}, term()} | ignore.
+initialize(#httpdb{} = HttpDb) ->
+    case remove_creds(HttpDb) of
+        {ok, User, Pass, HttpDb1} ->
+            case gen_server:start_link(?MODULE, [User, Pass, HttpDb1], []) of
+                {ok, Pid} ->
+                    {ok, HttpDb1, {Pid, 0}};
+                ignore ->
+                    ignore;
+                {error, Error} ->
+                    {error, Error}
+            end;
+        {error, missing_credentials} ->
+            ignore;
+        {error, Error} ->
+            {error, Error}
+    end.
+
+
+-spec update_headers(term(), headers()) -> {headers(), term()}.
+update_headers({Pid, Epoch}, Headers) ->
+    Args = {update_headers, Headers, Epoch},
+    {Headers1, Epoch1} = gen_server:call(Pid, Args, infinity),
+    {Headers1, {Pid, Epoch1}}.
+
+
+-spec handle_response(term(), code(), headers(), term()) ->
+    {continue | retry, term()}.
+handle_response({Pid, Epoch}, Code, Headers, Body) ->
+    Args =  {handle_response, Code, Headers, Body, Epoch},
+    {Retry, Epoch1} = gen_server:call(Pid, Args, infinity),
+    {Retry, {Pid, Epoch1}}.
+
+
+-spec cleanup(term()) -> ok.
+cleanup({Pid, _Epoch}) ->
+    gen_server:call(Pid, stop, infinity).
+
+
+%% Definitions
+
+-define(MIN_UPDATE_INTERVAL, 5).
+
+
+%% gen_server state
+
+-record(state, {
+    epoch = 0 :: non_neg_integer(),
+    cookie :: string() | undefined,
+    user :: string() | undefined,
+    pass :: string() | undefined,
+    httpdb_timeout :: integer(),
+    httpdb_pool :: pid(),
+    httpdb_ibrowse_options = [] :: list(),
+    session_url :: string(),
+    next_refresh = infinity :: infinity |  non_neg_integer(),
+    refresh_tstamp = 0 :: non_neg_integer()
+}).
+
+
+%% gen_server functions
+
+init([User, Pass, HttpDb]) ->
+    State = #state{
+        user = User,
+        pass = Pass,
+        session_url = get_session_url(HttpDb#httpdb.url),
+        httpdb_pool = HttpDb#httpdb.httpc_pool,
+        httpdb_timeout = HttpDb#httpdb.timeout,
+        httpdb_ibrowse_options = HttpDb#httpdb.ibrowse_options
+    },
+    case refresh(State) of
+        {ok, UpdatedState} ->
+            {ok, UpdatedState};
+        {error, {session_not_supported, _, _}} ->
+            ignore;
+        {error, Error} ->
+            {stop, Error}
+    end.
+
+
+terminate(_Reason, _State) ->
+    ok.
+
+
+handle_call({update_headers, Headers, _Epoch}, _From, State) ->
+    case maybe_refresh(State) of
+        {ok, State1} ->
+            Cookie = "AuthSession=" ++ State1#state.cookie,
```

Comment:
I am a bit concerned with this. Cookie value is not escaped and can be manipulated. I followed the code we use for `maybe_update_cookie` and it looks like there is a possibility to inject new headers.
```
CookieKVs = mochiweb_cookies:parse_cookie("AuthSession=\"see\nInjected=hack\";baz; Version=1").
Cookie = mochiweb_headers:get_value("AuthSession", CaseInsKVs).
40> "AuthSession=" ++ Cookie.
"AuthSession=see\nInjected=hack"
```

Reply:
But wondering what the threat model is here. The server somehow injecting extra headers into its own session cookie?